### ISSUE TITLE
How to implement a new item?

### ISSUE THREAD

**Issue by lucaszanella** [status-im/react-native-desktop#453]
I'm trying to implement a new item but I can't find, for example, where is a simple button implemented. I remember React Native for Android being very clear about how to create new items.

Is there a piece of c++ code that creates a QT button I can look into? 

I want to help :)

**Comment by vkjr** [comment]
Simple react native button implemented in pure js and has no counterpart on native side.
But in react-native-desktop you can find `NativeButton` component. Here is an implementation:
component manager:
https://github.com/status-im/react-native-desktop/blob/master/ReactQt/runtime/src/componentmanagers/buttonmanager.h
https://github.com/status-im/react-native-desktop/blob/master/ReactQt/runtime/src/componentmanagers/buttonmanager.cpp
Qml implementation:
https://github.com/status-im/react-native-desktop/blob/master/ReactQt/runtime/src/qml/ReactButton.qml
Js wrapper:
https://github.com/status-im/react-native-desktop/blob/master/Libraries/Components/NativeButton.desktop.js

**Comment by vkjr** [comment]
Here is a PR that contains `Switch` component implementation. It can give it a hint of all areas of code involved:
https://github.com/status-im/react-native-desktop/pull/305

**Comment by lucaszanella** [comment]
Thank you so much. Can you tell me why there are lots of objective-c files that looks like its from the core of the app and not just macOS implementation?

**Comment by lucaszanella** [comment]
As I see, the `Switch` component is mainly implemented in QML. I want to implement a something which is C++/QT in essence. Is the library  friendly for pure C++ implementations without even .ui components? Do you have an example of some sort?

Also, thanks, the `Switch` component helped me have a better view of how everything works :)

**Comment by vkjr** [comment]
@lucasfeliciano, yes, you can write c++ only implementation. Here is an example of extending `react-native-fs` library with code for desktop - https://github.com/status-im/react-native-fs/tree/master/desktop.

What kind of code do you want to create?

**Comment by lucaszanella** [comment]
I meant using QWidget instead of QQuickItem but apparently not. But ok, I'm already implementing my RTSP Video Viewer in QQuickItem. The idea is to build a Network Video Recorder in react native. 

Can you answer me just one more thing? Why do every module has an objective-c module counterpart? I see that QString `RNFSManager::moduleName()` returns its objective-c name and that objective-c file is from facebook's original implementation. Does that mean my new implemented item will have to implement some objective-c file? If so, what it should do?

**Comment by vkjr** [comment]
Currently `react-native-desktop` repository contains all the `react-native` code plus desktop-related code. This is why you see many objective c files in the project.
A bit later we will remove react-native code and react-native-desktop will be a separate project that installs along with react-native.

`react-native-fs` library is originally developed for `react-native`, so it contains native implementation for android and ios. I gave you a link to fork of this library within `status` organization. That fork extended with desktop-related code and this is why you see not only cpp, but also objective c code.

Within `status-react` repo we have few modules that implemented only for desktop, you can check them. For example - https://github.com/status-im/status-react/tree/develop/modules/react-native-desktop-menu

**Comment by lucaszanella** [comment]
While it's not possible to implement my item as a module, I'm trying to add it as a react native desktop component in the library. However, in the PR for switch https://github.com/status-im/react-native-desktop/pull/305 there's nothing that links it to the js libraries so I can import my item. Therefore I don't know how to give a name for my library in JS and how to import it. Are those the only files required or there's more?
